### ISSUE TITLE
3.0.3 - Hangs when using combination of flatMap and publish

### ISSUE THREAD

**Issue by sobersanta** [ReactiveX/RxJava#6974]
Hi.

After few days of desperate attempts to find the cause of my servers' spontaneous hangs, I managed to reproduce the issue, but still don't have a clear idea about the reason, but I think I have identified a possible direction.

All problems has started when I started using very promising operator groupBy with evicting map provider possibility:
```java
<K, V> Flowable<GroupedFlowable<K, V>> groupBy(@NonNull Function<? super T, ? extends K> keySelector,
            @NonNull Function<? super T, ? extends V> valueSelector,
            boolean delayError, int bufferSize,
            @NonNull Function<? super Consumer<Object>, ? extends Map<K, Object>> evictingMapFactory)
```

First I realised that it doesn't work well with Caffeine cache, as it performs evictions on a separate thread pool, and with async eviction notification groupBy goes crazy.

After moving to Guava things got better but only to some extent.

The tests I want to demonstrate I believe reproduce the issue I have. Definitely chain of operators is way more complicated in the production code, but here with certain variant of .publish() operator it's reproducible as well.

Basically, I'm not sure if there is issue in the groupBy(...evictingMapFactory) or in the publish(Function...), or in both.

Also tests are demonstrating that eviction policy doesn't work very well when crossing async boundaries downstream.

Please take a look at the attached file, it contains few simple tests, I kindly ask to check carefully all parameters, as it's important.

[FlowableOperators.java.zip](https://github.com/ReactiveX/RxJava/files/4604420/FlowableOperators.java.zip)

**Comment by akarnokd** [comment]
Hi and thanks for reporting this bug.

The problem is caused by the eviction double decrementing the group counter, once when the eviction completes the shared chain that cancels the group and once after the eviction.

I'll post a fix shortly.

**Comment by sobersanta** [comment]
Cool, looking forward for the fix!

**Comment by sobersanta** [comment]
Thanks for the prompt reply and the fix!
May I ask, when we can expect to have it to be released?
Maybe 3.0.4 soon? ;)

**Comment by akarnokd** [comment]
I suggest you verify it by testing with the current RxJava snapshot, just in case: https://github.com/ReactiveX/RxJava#snapshots

**Comment by sobersanta** [comment]
Sure, I will let you know about results.

**Comment by sobersanta** [comment]
Could you please comment on this?

```java
    public static void main(String[] args) {

        final int groups = 20;

        // Not completed (Timed out), buffer is too small
        int groupByBufferSize = groups * 2;
        int flatMapMaxConcurrency = 2 * groups;
        boolean notifyOnExplicitEviction = false;
        test2(groupByBufferSize, flatMapMaxConcurrency, groups, notifyOnExplicitEviction);

        // Success, buffer is large enough
        groupByBufferSize = groups * 30;
        flatMapMaxConcurrency = 2 * groups;
        test2(groupByBufferSize, flatMapMaxConcurrency, groups, notifyOnExplicitEviction);

        // Timeout... explicit eviction notification makes difference
        groupByBufferSize = groups * 30;
        flatMapMaxConcurrency = 2 * groups;
        notifyOnExplicitEviction = true;
        test2(groupByBufferSize, flatMapMaxConcurrency, groups, notifyOnExplicitEviction);

        // MissingBackpressureException, flatMap concurrency == groups count is not enough, must be at least 2x
        groupByBufferSize = groups * 20;
        flatMapMaxConcurrency = groups;
        notifyOnExplicitEviction = false;
        test2(groupByBufferSize, flatMapMaxConcurrency, groups, notifyOnExplicitEviction);
    }

    public static void test2(int groupByBufferSize, int flatMapMaxConcurrency, int groups,
                             boolean notifyOnExplicitEviction) {
        try {
            TestSubscriber<Integer> test = Flowable
                .range(1, 500_000)
                .map(i -> i % groups)
                .groupBy(i -> i, i -> i, false, groupByBufferSize,
                         // set cap too high
                         sizeCap(groups * 100, notifyOnExplicitEviction))
                .flatMap(gf -> gf
                             .take(10, TimeUnit.MILLISECONDS)
                    , flatMapMaxConcurrency)
                .test()
                .awaitDone(5, TimeUnit.SECONDS)
                .assertComplete();
            System.out.println("Success");
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

    public static <T> Function<Consumer<Object>, Map<T, Object>> sizeCap(int maxCapacity, boolean notifyOnExplicit) {
        return itemEvictConsumer ->
            CacheBuilder
                .newBuilder()
                .maximumSize(maxCapacity)
                .removalListener(n -> {
                    if (n.getCause() != RemovalCause.EXPLICIT || notifyOnExplicit) {
                        try {
                            itemEvictConsumer.accept(n.getValue());
                        } catch (Throwable throwable) {
                            throw new RuntimeException(throwable);
                        }
                    }
                })
                .<T, Object>build().asMap();
    }
```

**Comment by sobersanta** [comment]
changing `.take(10, MILLISECONDS)` to `.take(10)` makes only 3d case failing with timeout

**Comment by akarnokd** [comment]
Is this with the snapshot?

**Comment by sobersanta** [comment]
yes, snapshot

**Comment by sobersanta** [comment]
`gf.observeOn(Schedulers.computation()).take(10)` produces yet another set of results.

**Comment by akarnokd** [comment]
I don't have the time today to look into these but we can perhaps narrow it down. Do the hangs happen without the eviction customization?

**Comment by sobersanta** [comment]
Behaviour is slightly different, but yes, on 3.0.3 also.

**Comment by sobersanta** [comment]
I don't know if this can be in help, but so far I got to the following:
synchronizing `GroupBySubscriber.onNext` with `State.cancel` fixes the issue.

```java

public synchronized void onNext(T t) {
////....

// ... State...

@Override
        public void cancel() {
            synchronized(parent) {
                if (cancelled.compareAndSet(false, true)) {
                    cancelParent();
                    drain();
                }
            }
        }
```

**Comment by akarnokd** [comment]
Further fixes in #6979.

**Comment by sobersanta** [comment]
OK it seems to me I got the core of the issue. It lies in the `.drain()` method and the backpressure coordination with the upstream.
If group cancel happens concurrently with parent's onNext call, it's possible that group's onNext is called after cancel.
In this case, drain will return immediately as work-in-progress atomic guard is not updated when calling `checkTerminated`.
So that call is not accounted in the backpressure coordination inside `checkTerminated` and leads to a hang.
Easiest brute-force way to get rid of this problem is to change here return->break:
```java
if (checkTerminated(d, empty, a, delayError, e)) {
    return; // <- here should be `break`,
}
```

But I believe there should be more correct solution.

Also I think another valuable change would be to add verification that the group being removed in `.cancel(key)` is exactly the group that is calling this method, not blindly removing the key, as it theoretically may contain another group already, created after e.g. group abandonment or map eviction.
Schematically:
```java

        public void cancel(GroupedUnicast<K, V> group) {
            K key = group.getKey();
            Object mapKey = key != null ? key : NULL_KEY;
            if (groups.remove(mapKey, group)) {
                if (groupCount.decrementAndGet() == 0) {
                    upstream.cancel();
                }
            } else {
                System.out.println("Group doesn't exist anymore");
            }
        }

/// ... in State


        State(int bufferSize, GroupBySubscriber<?, K, T> parent, K key, boolean delayError, GroupedUnicast<K, T> group) {
            this.queue = new SpscLinkedArrayQueue<>(bufferSize);
            this.parent = parent;
            this.key = key;
            this.delayError = delayError;
            this.group = group;
        }


        void cancelParent() {
            if ((once.get() & ABANDONED) == 0) {
                parent.cancel(group);
            }
        }

**Comment by sobersanta** [comment]
And another thing which confuses me:

in drainNormal there is a call to q.poll() before a checkTerminated(), which removes element from the queue, and it's not being accounted in the coordination.

**Comment by sobersanta** [comment]
OK I see you already get your hands to the issue :)
I can see you have covered the problems,
and the only possible issue I still can see is the blind group key removal.
Correct me if I'm wrong:
if there is a race between evicting group from cache by the cache itself, calling parent's onNext and cancellation of the group,
it's possible that the cache removal will happen, then new group is created in onnext, and then cancellation happens and also removes the group from the cache, but there is already a new group, so it's an extra group removal.

**Comment by akarnokd** [comment]
> issue I still can see is the blind group key removal

Could be. A test case would be great for this.

> `groups.remove(mapKey, group)`

This could be a problem. The API uses `Map` and `remove(K, V)` was added to it in Java 8. I don't know how Android Desugar would handle it. Currently, all pre-Java 8 operators are still restricted to use Java 6 API just in case. Also [`ConcurrentMap.remove(K, V)`](https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/ConcurrentMap.html#remove(java.lang.Object,%20java.lang.Object)) does exist but would constitute as a public API change (and perhaps a limiting factor for existing usages).

**Comment by sobersanta** [comment]
Not sure how to write a test case for that... It shouldn't lead to a hang, just unnecessary group removal.

**Comment by sobersanta** [comment]
As you mentioned compatibility with Android and older Java runtimes, may I ask the question which is not related to the issue...
Is that compatibility restriction a reason for not using AtomicFieldUpdaters in the library in favour of pointer-chasing Atomics?

**Comment by akarnokd** [comment]
Reflective field access is broken on some Android phones.

Field updaters/varhandles are not always a clear win due to false sharing introduced by jam-packed state. You'd need to pad fields which increases the memory usage for all cases, not just the high-performance scenarios. Also on latest JDKs, there are signs that the previously widely employed class-hierarchy-based padding trick no longer works.

**Comment by sobersanta** [comment]
More than clear, thanks for the answer!

**Comment by sobersanta** [comment]
As for the Map's remove(Object,Object).
WDYT about checking if the map is instance of ConcurrentMap (I believe it will be like that in most of the cases, I can hardly imagine anyone would uses anything else), and if yes then using ConcurrentMap.remove(key, value), otherwise keep existing logic?
That check can be done during the construction, and for example keep 2 nullable references to maps, or wrap a non-ConcurrentMap's implementation with a proper delegating wrapper.

**Comment by sobersanta** [comment]
I have tried the code from your latest PR. And my pipeline still hangs on the graceful completion :(
So far don't have a simple test which is reproducing the problem, will let you know when I get something. Could be a problem is in another operator. (Though it works fine if I serialize calls to State.cancel and upstream onNext in groupby)

**Comment by sobersanta** [comment]
OK here is the case (just added `.observeOn` to grouped flowable):

```java

public static void test2(int groupByBufferSize, int flatMapMaxConcurrency, int groups,
                             boolean notifyOnExplicitEviction) {
        try {
            Flowable
                .range(1, 500_000)
                .map(i -> i % groups)
                .doOnCancel(() -> System.out.println("Cancelling upstream"))
                .compose(s -> groupBy(s, i -> i, i -> i, groupByBufferSize, false,
                                      sizeCap(groups *2, notifyOnExplicitEviction)))
                .flatMap(gf -> gf
                             .observeOn(Schedulers.computation())
                             // .take(10)
                             .take(10, TimeUnit.MILLISECONDS)
                    , flatMapMaxConcurrency)
                .test()
                .awaitDone(5, TimeUnit.SECONDS)
                .assertComplete();
            System.out.println("Success");
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }
```

**Comment by akarnokd** [comment]
If you put `hide()` before `observeOn`, does it work then?

**Comment by sobersanta** [comment]
> If you put hide() before observeOn, does it work then?

Apparently yes

**Comment by sobersanta** [comment]
I can see that partially my problems are caused by the .publish() operator (it uses FlowablePublishMulticast) which on complete calls .dispose() which in turn cancels upstream. I'm not sure I understand the reason of that canceling while this is a normal graceful shutdown. Could you please elaborate a little for a dummy like me why is it cancelling upstream on complete event? It messes with my resource cleanup logic (onTerminate is never called after that), and I'm not sure how to overrun it.

BTW here's the case where the new code fails even with `.hide()` in place:

```java

    public static void test2(int groupByBufferSize, int flatMapMaxConcurrency, int groups,
                             boolean notifyOnExplicitEviction) {
        try {
            Flowable
                .range(1, 500_000)
                .map(i -> i % groups)
                .doOnCancel(() -> System.out.println("Cancelling upstream"))
                .compose(s -> groupBy(s, i -> i, i -> i, groupByBufferSize, false,
                                      sizeAndTtlCapGuava(groups * 2, Duration.ofMillis(10), notifyOnExplicitEviction)))
                .flatMap(gf -> gf
                             .hide()
                             .publish(shared -> shared)
                    , flatMapMaxConcurrency)
                .test()
                .awaitDone(5, TimeUnit.SECONDS)
                .assertComplete();
            System.out.println("Success");
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

    public static <T> Function<Consumer<Object>, ConcurrentMap<T, Object>> sizeAndTtlCapGuava(int maxCapacity,
                                                                                              Duration ttl,
                                                                                              boolean notifyOnExplicit) {
        return itemEvictConsumer ->
            CacheBuilder
                .newBuilder()
                .maximumSize(maxCapacity)
                .expireAfterWrite(ttl)
                .removalListener(n -> {
                    if (n.getCause() != com.google.common.cache.RemovalCause.EXPLICIT || notifyOnExplicit) {
                        try {
                            itemEvictConsumer.accept(n.getValue());
                        } catch (Throwable throwable) {
                            throw new RuntimeException(throwable);
                        }
                    }
                })
                .<T, Object>build().asMap();
    }

```

**Comment by akarnokd** [comment]
> why is it cancelling upstream on complete event

With `publish(Function)`, the inner flow may be independent of the upstream so when the inner terminates, the upstream should be stopped. I suppose when there is a causal chain from an upstream completion to a downstream completion, this cancellation could be suppressed.

I'll look into the new cases tomorrow.

**Comment by sobersanta** [comment]
Adding .observeOn leads to MissingbackpressureException:

```java

    public static void main(String[] args) {
        final int groups = 20;

        // MissingBackpressureException
        int groupByBufferSize = 2;
        int flatMapMaxConcurrency = 2 * groups;
        boolean notifyOnExplicitEviction = false;
        test2(groupByBufferSize, flatMapMaxConcurrency, groups, notifyOnExplicitEviction);

        groupByBufferSize = groups * 30;
        flatMapMaxConcurrency = 2 * groups;
        test2(groupByBufferSize, flatMapMaxConcurrency, groups, notifyOnExplicitEviction);
    }

    public static void test2(int groupByBufferSize, int flatMapMaxConcurrency, int groups,
                             boolean notifyOnExplicitEviction) {
        try {
            Flowable
                .range(1, 500_000)
                .map(i -> i % groups)
                .compose(s -> groupBy(s, i -> i, i -> i, groupByBufferSize, false,
                                      sizeAndTtlCapGuava(groups, Duration.ofMillis(10), notifyOnExplicitEviction)))
                .flatMap(gf -> FlowableOperators.publish(gf.hide()
                                                           .observeOn(Schedulers.computation())
                    , shared -> shared)
                    , flatMapMaxConcurrency)
                .test()
                .awaitDone(5, TimeUnit.SECONDS)
                .assertComplete();
            System.out.println("Success");
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }
```

**Comment by sobersanta** [comment]
Continuing about problem with the latest test above.
What I realised today is that when a group completes (e.g. because of external eviction from the Map), it pushes down onComplete event, and in case there is observeOn operator downstream, or any other reason which causes onComplete to be scheduled on another thread, the consequence is that expected group request from flatMap (or another operator consuming groups) happens with a delay, which leads to "group abandonment". It's not really about subscribing to a new group only, as stated in the documentation, but requesting more when another one completes.

Simple reproduction:

```java
Flowable
  .range(1, 500_000)
  .map(i -> i % 20)
  .groupBy(i->i) // no any group eviction
  .flatMap(gf -> gf.hide()
                                 .take(10)
                                 .observeOn(Schedulers.computation())
                    ,1000)
  .test()
  .awaitDone(5, TimeUnit.SECONDS)
  .assertComplete();
```

Concurrency in flatMap of 1000 is not enough

**Comment by sobersanta** [comment]
And when group is cancelled e.g. by take() operator, and it's done asynchronously, this is again is not accounted into the pair "requested-emitted" and leads to group abandonment.
I believe number of completed/cancelled groups should be accounted in the abandonment error generation.

**Comment by akarnokd** [comment]
Are you sure you get group abandonment instead of MissingBackpressureException? Abandonment happens only when there is no synchronous subscription happening but there were requests for new groups.

Also if you delay the completion within flatMap, there is no way for the operator to request more until there are concurrency slots freeing up.

**Comment by sobersanta** [comment]
Maybe I’m calling it wrong... It’s naturally being delayed: you evict group
from the map, immediately schedule group eviction task which will in turn
call onComplete. But that will happen eventually, along with following
request from flatMap.
But in that turn of onNext a new group is already trying to be emitted
which is causing error as requests didn’t catch up with emits.

**Comment by sobersanta** [comment]
Basically - isn’t it a violation of the protocol? we shouldn’t emit anything
unless requested. I think such eviction should be coordinated with the
emission? Conceptually we have just completed previous group ( in Guava
evictions happen exactly when you call get/put) we are allowed to emit new
group, but only after it completes and flatMap requests more. So maybe it
should queue up such emissions until next request from downstream?

**Comment by akarnokd** [comment]
I've experimented with the new cases and here is the rundown:

- `gf.publish(shared->shared)` does not hang for me.
- `gf.observeOn().take()` hangs but `gf.hide().observeOn().take()` passes.

The latter could lead down the rabbit hole so I'll merge #6979 to have a clean baseline for further actions. Let's continue the discussion in #6982.
